core/src/impl/Kokkos_FunctorAnalysis.hpp

```diff
@@ -625,29 +637,41 @@ struct FunctorAnalysis {
     }
   };
 
+  template <class F>
+  struct DeduceJoinNoTag<F, std::enable_if_t<(is_reducer<F>::value ||
+                                              (!is_reducer<F>::value &&
+                                               std::is_void<Tag>::value)) &&
+                                             detected_join_no_tag<F>::value>>
+      : public has_join_no_tag_function<F> {
+    enum : bool { value = true };
+  };
+
   template <class F>
   struct DeduceJoinNoTag<
       F,
       std::enable_if_t<(is_reducer<F>::value ||
                         (!is_reducer<F>::value && std::is_void<Tag>::value)) &&
-                       (detected_join_no_tag<F>::value ||
+                       (!detected_join_no_tag<F>::value &&
                         detected_volatile_join_no_tag<F>::value)>>
-      : public std::conditional_t<detected_join_no_tag<F>::value,
-                                  has_join_no_tag_function<F>,
-                                  has_volatile_join_no_tag_function<F>> {
+      : public has_volatile_join_no_tag_function<F> {
```

Comment:
Do we don't warn if there is both a `volatile`-qualified version and a non-`volatile`-qualified version? I think I would prefer to also warn in that case but I don't feel strongly about it.

Reply:
In that case, the warning would only be suggesting code clean-up, not performance pitfall avoidance. The code could be that way intentionally to maintain compatibility with Kokkos <3.7. We'd also need an additional case of the SFINAE logic, I think?

I think that all leans against trying to warn in that case.

Reply:
we can consider that later again, its not important for the release.